### DISCUSSION TITLE
Cross compiling for macOS ARM64 producing x86 libraries (cmake involved)

### DISCUSSION THREAD

**Discussion by epeisach** [pypa/cibuildwheel#1706] [Q&A]
`export CIBW_ARCHS_MACOS="x86_64 arm64"
             cibuildwheel --config cibuildwheel.toml --output-dir "$(System.DefaultWorkingDirectory)/dist" .`
cibuildwheel.toml:
`[[tool.cibuildwheel.overrides]]
select = "*-musllinux*"
before-all = "apk add flex bison cmake"`

So setup.py is invoking cmake.  The built libraries for macOS are MachO - x86_64 in both case.

I suspect that I will need to pass a flag to cmake to enable cross compiling.   
a) Is there an environment variable set indicating building architecture - so I can detect in setup.py
b) Any ideas what I would pass to CMAKE?

**Comment by henryiii** [comment]
I would recommend using [scikit-build-core](https://scikit-build-core.readthedocs.io/en/latest/) or other dedicated backend. If you want to do it yourself, see https://github.com/pybind/cmake_example/blob/cbcf742c566a8bfd1f079b3d87778b17fef4bef5/setup.py#L103/

**Reply by epeisach** [reply]
Thank you ---- this is a pybind scenario.  I suppose I can set an environment variable for cibuildwheel - so only attach architecture if cibuildwheel.  That way distinguish cibuildwheel from someone building naively.

**Reply by henryiii** [reply]
No, you want to solve this for all possible builders, like the conda family too. You should read ARCHFLAGS and convert it to `CMAKE_OSX_ARCHITECTURES`, like I showed in the example above.

Though I'd highly recommend a dedicated builder over handling every possible situation yourself. More things will arise that scikit-build-core has already solved.

**Comment by epeisach** [comment]
Do to requirement for python 2 support - at the moment, I went with the example - and it works for now.  I have another project that I will retrofit to use scikit-build-core.
